File: src/main/java/org/hildan/minecraft/mining/optimizer/patterns/BranchingPattern.java
package org.hildan.minecraft.mining.optimizer.patterns;

import org.hildan.minecraft.mining.optimizer.blocks.Sample;
import org.hildan.minecraft.mining.optimizer.geometry.Axis;
import org.hildan.minecraft.mining.optimizer.patterns.tunnels.TunnelPattern;

import java.util.HashSet;
import java.util.Set;

/**
 * One main shaft with perpendicular branches.
 */
public class BranchingPattern extends AbstractDiggingPattern {

    private final TunnelPattern shaft;

    private final TunnelPattern branch;

    private final int branchLength;

    private final int branchOffsetByTier;

    public BranchingPattern(TunnelPattern shaft, TunnelPattern branch, int branchLength, int branchOffsetByTier) {
        this.shaft = shaft;
        this.branch = branch;
        this.branchLength = branchLength;
        this.branchOffsetByTier = branchOffsetByTier;
        if (shaft.getShape().getHeight() < branch.getShape().getHeight()) {
            throw new IllegalArgumentException("The main shaft should be higher than branches");
        }
        if (shaft.getHSpacing() < 2 * branch.getShape().getWidth()) {
            throw new IllegalArgumentException(
                    "Branches from 2 different shafts are touching: reduce branch length, or put more space");
        }
    }

    @Override
    public int getWidth() {
        // the offset doesn't matter here, the spatial period is the same
        return 2 * branchLength + shaft.getShape().getWidth();
    }

    private int getLayerHeight() {
        return shaft.getShape().getHeight() + shaft.getVSpacing();
    }

    @Override
    public int getHeight() {
        // with an offset, two consecutive layers are different
        return getLayerHeight() * 2;
    }

    @Override
    public int getLength() {
        return branch.getHSpacing() + branch.getShape().getWidth();
    }

    @Override
    public Set<Access> getAccesses(int x, int y) {
        Set<Access> accesses = new HashSet<>();
        accesses.add(new Access(x + branchLength, y));
        accesses.add(new Access(x + branchLength, y + getLayerHeight()));
        return accesses;
    }

    @Override
    public void digInto(Sample sample, int originX, int originY, int originZ) {
        digLayer(sample, originX, originY, originZ, 0);
        digLayer(sample, originX, originY + getLayerHeight(), originZ, branchOffsetByTier);
    }

    private void digLayer(Sample sample, int originX, int originY, int originZ, int offset) {
        branch.digInto(sample, originX, originY, originZ + offset, branchLength, Axis.X, Axis.Y);
        shaft.digInto(sample, originX + branchLength, originY, originZ, getLength(), Axis.Z, Axis.Y);
        int oppositeBranchStartX = originX + branchLength + shaft.getShape().getWidth();
        branch.digInto(sample, oppositeBranchStartX, originY, originZ + offset, branchLength, Axis.X, Axis.Y);
    }
}

File: src/main/kotlin/org/hildan/minecraft/mining/optimizer/patterns/tunnels/BranchingPattern.kt
package org.hildan.minecraft.mining.optimizer.patterns.tunnels

import org.hildan.minecraft.mining.optimizer.blocks.Sample
import org.hildan.minecraft.mining.optimizer.patterns.Access
import org.hildan.minecraft.mining.optimizer.patterns.RepeatedDiggingPattern

/**
 * One main shaft with perpendicular branches.
 */
class BranchingPattern(
    private val shaft: TunnelPattern,
    private val branch: TunnelPattern,
    private val branchLength: Int,
    private val branchOffsetByTier: Int,
) : RepeatedDiggingPattern {

    private val layerHeight = shaft.section.height + shaft.vSpacing

    // the offset doesn't matter here, the spatial period is the same
    override val width = 2 * branchLength + shaft.section.width

    // with an offset, two consecutive layers are different
    override val height = layerHeight * 2

    override val length = branch.hSpacing + branch.section.width

    init {
        if (shaft.section.height < branch.section.height) {
            throw IllegalArgumentException("The main shaft should be higher than branches")
        }
        if (shaft.hSpacing < 2 * branch.section.width) {
            throw IllegalArgumentException("Branches from 2 different shafts are touching: reduce branch length, or put more space")
        }
    }

    override fun getAccesses(offsetX: Int, offsetY: Int) =
        setOf(Access(offsetX + branchLength, offsetY), Access(offsetX + branchLength, offsetY + layerHeight))

    override fun digInto(sample: Sample, offsetX: Int, offsetY: Int, offsetZ: Int) {
        digLayer(sample, offsetX, offsetY, offsetZ, 0)
        digLayer(sample, offsetX, offsetY + layerHeight, offsetZ, branchOffsetByTier)
    }

    private fun digLayer(sample: Sample, originX: Int, originY: Int, originZ: Int, offset: Int) {
        branch.digInto(sample, originX, originY, originZ + offset, branchLength, Axis.X)
        shaft.digInto(sample, originX + branchLength, originY, originZ, length, Axis.Z)
        val oppositeBranchStartX = originX + branchLength + shaft.section.width
        branch.digInto(sample, oppositeBranchStartX, originY, originZ + offset, branchLength, Axis.X)
    }
}
